### ISSUE TITLE
Fledge out the design for messages vs Tx

### ISSUE THREAD

**Issue by liamsi** [lazyledger/lazyledger-core#67]
## Summary

Specify a way to distinguish between `Tx` and `Messages` in ll-core / abci. 
From the point of view of lazyledger-core: Tx = opaque blobs that get executed by the abci app (like Tx in tendermint), and, Messages = blobs with no impact on the state. 
While both (Messages and Tx) have to be part of the `Block.Data`, `Tx` do not explicitly carry a namespace ID (the reserved ID for Tx is only added when computing the shares to not add redundant data) but `Messages` essentially include their own namespace ID (e.g. by prefixing).

## Problem Definition

To be able to compute and commit the namespaced shares properly to the namespaced merkle tree, lazyledger-core needs a way to understand the difference between blob that will have a reserved namespace and blob that provides it's own namespace. 

Although messages do not change state they also need to be present at the abci-app level: a tx is only valid if the message that it is pointing to is present (and the commitment in the Tx matches the message).

## Proposal

Several ideas and questions arose that we need clarification. Here is an incomplete list:

- we use already precomputed shares as (tendermint) Tx. Major downside: It would increase the size of Block.Data unnecessarily and it is very different from the spec. My gut feeling is that it might also be impractical due to other reasons. 
- can we use some of the existing abci methods to sneak in that distinction?
- clarify if preprocessing the Tx block could be helpful here (ref #59), or, DeliverBlock?
- extend or modify the abci interface (ideally in a way that requires little or no modification to the SDK modules)

Another thing we need to think about is if we want to keep the messages in the same (mem)pool as Tx (that get executed).

For some context of the mempool and it's connection to the app: 
The mempool reactor gets a blob, it decodes it [here](https://github.com/lazyledger/lazyledger-core/blob/7b84a4c74317453c6fd4192b1b82619328b1b97c/mempool/reactor.go#L165) and then asks the app (via CheckTx) to include the tx or not [here](https://github.com/lazyledger/lazyledger-core/blob/7b84a4c74317453c6fd4192b1b82619328b1b97c/mempool/reactor.go#L177). 

Required reading:
 - https://docs.tendermint.com/master/spec/abci/abci.html
 - https://docs.cosmos.network/master/basics/tx-lifecycle.html

**Comment by liamsi** [comment]
it's probably not a bad idea to think about what should be the Tx that lazyledger-core includes in the block? Currently, in tendermint they are just opaque blobs (Tx) with no inherent structure, besides what protobuf imposes (they are wrapped in a proto message - only the raw Tx end up in the block though): https://github.com/lazyledger/lazyledger-core/blob/7b84a4c74317453c6fd4192b1b82619328b1b97c/proto/tendermint/mempool/types.proto#L6-L14

So far we roughly discussed 4 ideas:
 - Tx are opaque blobs -> that means only the app knowns how to extract the namespace ID (and hence computation of the data root using the nmt will need to happen in the app)
- Txs from the point of view tendermint are pairs of Tx and Messages 
- (similar to above) pair of Tx and optionally (namespaced) shares; only for [`PayForMessage`](https://github.com/lazyledger/lazyledger-specs/blob/master/specs/data_structures.md#signedtransactiondata-payformessage) transactions there will be shares present 
- all Tx are messages -> change all Tx to contain a namespace ID (or implicitly assume the reserved ID for Tx if it isn't present)

**Comment by liamsi** [comment]
To roughly visualize the above, I've summarized these 4 options in [this pdf](https://github.com/lazyledger/lazyledger-core/files/5408034/Tx-Messages.4.pdf).

To summarize our discussions and the current conclusions: it seems like the best approach to leave Tx to be just blobs from the PoV of tendermint. At least to the point where they are checked against the app via CechkTx.
That means the Tx include their message (so both can be checked if they should enter the mem-pool and gossiped around as a single entity). This means there must be way to extract the message from the Tx though. How this could look like becomes clearer if we look at how Tx that are submitted should look like and to understand this best, I try to describe the lifecycle of a (LL) transaction from creation to inclusion in the mempool and later in the block:

Note: below we mostly consider Tx that pay for the inclusion of messages

### Sign-Data

Before signing the Tx, the client would need to compute potential share layouts, split the message it wants to include
into corresponding shares and commit to those.
The client then signs multiple such Tx (for each relevant share layout depending on potential square-sizes).

More concretely, signatures are computed on objects that look - from the PoV of the app's client -
sth like this (before encoding obviously):
```proto
message {
 // ... non-LL specific fields skipped
 bytes  nid   =  1;
 bytes  commit = 2; // commit(msg-shares)
};
```

### Tx

The Tx that gets sent to the mem-pool is one Tx, before encoding, looks like this:

```proto
message {
 // ... non-LL specific fields skipped
 bytes  nid   =  1;
 bytes  msg-data = 2; // the actual message this Tx pays for (not split into shares)
 repeated bytes signatures = 3; // signatures of above's sign-data
};
```

Note that, after encoding, this remains an opaque byte blob from the PoV of tendermint.
As mentioned above this is how the Tx could be passed to the mem-pool and how the Tx is
gossiped between nodes (as an opaque blob).

### Further steps and processing

To verify if a Tx should enter the mem-pool, tendermint asks the app via CheckTx.
We noted that to decide if a Tx should enter the pool, the app (like the client above)
would already need to
compute shares (corresponding to potential layouts) to verify the signatures in the Tx.

Hence, one option could be for CheckTx to already return multiple Tx (one for each sig) and
the corresponding (single) Message Data. That would mean some more changes to the mem-pool:
basically the pool would need to track multiple "things", and their relation,
instead of one blob.
This needs further investigation!

Alternatively, CheckTx would at least need to return IDs that can be used to [evict the pool](https://github.com/lazyledger/lazyledger-core/issues/81).

The latter means that we could instead use a pre-process phase (because we can use these IDs
to evict the mem-pool after a block with the processed Tx was seen).

### Further required clarifications

It's not yet entirely clear yet how in the above drafted approach,
the square-size is [decided](https://github.com/lazyledger/lazyledger-core/issues/77) and if that would 'live' entirely on the app-side of abci, or,
in tendermint.

**Comment by liamsi** [comment]
IMO, we can close this. Following discussions that emerged from the implementation, the spec now sufficiently describes Tx vs Messages. The remaining gaps are filled via what the usage of tendermint and the cosmos-sdk impose anyways.

There are still some considerations to be made here:
 - reaping from the mempool depending on max square / block size (#77) 
 - making sure the mempool can actually deal with large tx (the combo of tx+message+signatures can become quite large - not sure the current logic will reach its limits with this quickly)

There might be more to consider. Either way it seems like we should rather open dedicated issues to resolve any remaining issues.

If you disagree, feel free to reopen this one!
